Mage/src/main/java/mage/abilities/effects/common/SacrificeSourceEffect.java

```diff
@@ -40,7 +48,7 @@ public boolean apply(Game game, Ability source) {
         if (sourceObject instanceof Permanent) {
             Permanent permanent = (Permanent) sourceObject;
             // you can only sacrifice a permanent you control
```

Comment:
Looks like no needs in controller check at all — permanent must be sacrificed anyway. There are not controller check in all other usages of sacrifice.

Reply:
No, there must be a check for the controller for most effects.

701.17a. To sacrifice a permanent, its controller moves it from the battlefield directly to its owner's graveyard. **A player can't sacrifice something that isn't a permanent, or something that's a permanent they don't control.** Sacrificing a permanent doesn't destroy it, so regeneration or other effects that replace destruction can't affect this action.

Reply:
Can you describe real cards example with canceled sacrifice due that rule?

Reply:
[[Ball Lightning]]. With the trigger on the stack, cast [[Wrong Turn]] to donate it to an opponent. The Ball Lightning must survive when the trigger resolves, since you control the trigger but not the creature.

Reply:
Interesting. Another example: https://www.reddit.com/r/mtgrules/comments/186z1p1/mishrass_warform_delayed_sacrifice_trigger_if/

> The delayed trigger of "Sacrifice it at the beginning of the next end step" is a result of the resolving trigger from Mishra. (CR603.7)
> Since you controlled the original trigger, you control the delayed trigger. (CR603.7e)
> You can't sacrifice something you don't control (CR701.17a), so when the delayed trigger resolves, nothing happens.
> Contrast to [[Mizzix, Replica Rider]], which grants the sacrifice trigger to the creature itself, so whoever controls the creature when the trigger goes on the stack sacrifices it.

Result depends on source ability:
- normal ability;
- triggered ability;
- delayed triggered ability;

Looks like sacrifice usage must be research for possible bugs. It can be potentially a massive problem.

Reply:
“source.isControlledBy(permanent.getControllerId()” — that’s conditional must process all use cases.

Maybe it can be moved sacrifice code:
![IMG_0620](https://github.com/user-attachments/assets/f4b3f35f-834d-492b-ab63-67fcd5638586)

Reply:
> controllerSacrifices

Maybe it can be implemented as sacrifice param.

@ssk97 can you check that cards implementation — is it affected?

https://scryfall.com/search?q=o%3A%22controller+sacrifice%22

Reply:
Right now the `sacrifice` function implicitly assumes that the controller is sacrificing it, which is usually the correct behavior. The exceptions are here in `SacrificeSourceEffect` ("Sacrifice ~") and in `SacrificeTargetEffect` with the `playerIdThatHasToSacrifice` constructor used for cards like Gemini Engine.

I looked through that list and as far as I can tell all of the cards are already correctly implemented.